### PR TITLE
Added four preview images

### LINKED ISSUE #247: grid of thumbnails showing step options in addStep ui
We could generate a tiny thumbnail of the final image and run 5 or so of the most popular or useful modules, showing very low res previews of what they'd look like. Sort of like in Instagram!

(Low priority but fun, also dependent on a scale module)

### PR DESCRIPTION
**Pull request by okonek**
Fixes #247 [Add issue number here.]

Make sure these boxes are checked before your pull request (PR) is ready to be reviewed and merged. Thanks!

* [x] tests pass -- look for a green checkbox ✔️ a few minutes after opening your PR -- or run tests locally with `rake test`
* [x] code is in uniquely-named feature branch and has no merge conflicts
* [x] PR is descriptively titled
* [x] ask `@publiclab/reviewers` for help, in a comment below

> We're happy to help you get this ready -- don't be afraid to ask for help, and **don't be discouraged** if your tests fail at first!

If tests do fail, click on the red `X` to learn why by reading the logs.

Please be sure you've reviewed our contribution guidelines at https://publiclab.org/contributing-to-public-lab-software

Thanks!

thumbnails

### CODE REVIEW

**Comment by okonek** [comment]
@jywarren Is everything ok in here?

**Comment by okonek** [comment]
@publiclab/reviewers

**Comment by kevinzluo** [comment]
Hi @okonek . Your implementation looks pretty cool!

Just as a suggestion, could we maybe display each image preview under its respective button and maybe allow them to click it?

image

**Comment by okonek** [comment]
Sure, that's a great idea. I'll do it.

**Comment by jywarren** [comment]
Ooh this is awesome. I love kevinzluo's suggestion too! Perhaps we could
even show the icon in white overlaid on the preview image?

**Comment by jywarren** [comment]
Hi @okonek -- it looks like @JonathanXu1 was working on this already in #504 -- did you have a chance to coordinate with Jonathan? Perhaps there is some collaboration that could happen?

**Comment by jywarren** [comment]
For example i know @JonathanXu1 developed some nice mockups for different projects and maybe could contribute some design ideas or even CSS code?

**Comment by JonathanXu1** [comment]
I'd love to collaborate on this :)

**Comment by okonek** [comment]
@JonathanXu1 @jywarren 
Ok, so I've added this feature to show thumbnails on buttons, but there is a more stylish problem with the resize thumbnails, because I don't know how to crop it to keep the properties of width and height. Maybe you could help on this one.
thumbnails2

**Comment by okonek** [comment]
@JonathanXu1 Well, thank you very much, but I don't know if there is anything more to do in there. However if you have any idea of how it could be done better, let me know. I'd be very happy to collaborate with someone.

**Comment by okonek** [comment]
@jywarren What do you think about this solution? Is it acceptable for you?

**Comment by jywarren** [comment]
How about something like this? (i've used stand-in icons and images but you get the idea)

image

@JonathanXu1 if you have suggestions that's awesome. @okonek in the future please just ask in a comment on the original issue to be sure someone's not already started -- even if they have, perhaps you can help them solve it! 

Thanks everyone!

**Comment by JonathanXu1** [comment]
@jywarren @okonek I can help design the preview like that. Since I've been working on this issue over the past week, I'd really appreciate the opportunity to finish it.

**Comment by jywarren** [comment]
How about this: @JonathanXu1 can do some html/css design of the interface, but building on the code in this PR -- and then we merge the joint work. How does that sound? Also, just a shout out to @kevinzluo for the great design input as well! Good collaboration, folks! 🙌

**Comment by JonathanXu1** [comment]
Great idea. Is there a way to clone code from a pull request?

**Comment by okonek** [comment]
I'm very sorry for taking your task, but I really didn't know. Maybe if someone has took the task like that it should be removed from GCI tasks list?

**Comment by okonek** [comment]
@JonathanXu1 This could be helpful in creating a fork from a fork https://help.github.com/articles/creating-a-pull-request-from-a-fork/ Also remember that the code is on the branch `add-grid-of-thumbnails`
Good luck!

**Comment by jywarren** [comment]
@okonek is right about PR from a fork - the cool part about that will be that once @JonathanXu1's code is ready, we can merge it and it'll then appear in this PR and the tests will be re-run. Very meta! 🐢🐢🐢 

@okonek no worries, it happens pretty often that folks don't see one another, as we're spread all over the world and the code is spread all over different parts of GitHub 😄  it's a lot to keep track of!

I'm going to approve @okonek's portion of the task but make a new task instance for the UI part if @JonathanXu1 wants to take that up. It's your option!

Thanks, all!

**Comment by tech4GT** [comment]
This UI works looks really good!! Nice work @okonek :clap:

**Comment by JonathanXu1** [comment]
How's this for the preview styling?
image

**Comment by jywarren** [comment]
It's great!!!! Could we actually constrain these to be square, using some
CSS? We might need to wrap them in a div and use `overflow:hidden`, make
the div square, and put the image inside it, and ensure the smallest
dimension of the image (height or width) is at least as large as the
container div?

**Comment by okonek** [comment]
@jywarren Thank you!

**Comment by kevinzluo** [comment]
Hello! This is looking great!

I think this task might have just been duplicated though. Look here: https://codein.withgoogle.com/tasks/6720201132343296/?sp-organization=6519545549291520&sp-is_beginner=False.

**Comment by JonathanXu1** [comment]
How is this? @jywarren @okonek 
image
Also I believe there are two instances of the task because Okonek and I are collaborating on this issue.

**Comment by tech4GT** [comment]
This looks amazing!! Thanks a lot @JonathanXu1

File: src/modules/Crop/Module.js [hunk 1]
```diff
@@ -17,7 +17,7 @@ module.exports = function CropModule(options, UI) {
 
   // we should get UI to return the image thumbnail so we can attach our own UI extensions
   // add our custom in-module html ui:
-  if (options.step.inBrowser) var ui = require('./Ui.js')(options.step, UI);
+  if (options.step.inBrowser && !options.noUI) var ui = require('./Ui.js')(options.step, UI);
```
[the remark this answers is not in the document]
Comment (tech4GT) [reply src/modules/Crop/Module.js]: Oh great that makes sense! I'll try this out in my browser and then we can merge! :blush:

**Comment by tech4GT** [comment]
Could you also update the dist files here, just run `grunt build` and commit the new dist files. :smile:

**Comment by JonathanXu1** [comment]
Thanks! I'd like to do that but I don't have permissions to modify files
image

**Comment by JonathanXu1** [comment]
I've made pull request https://github.com/jywarren/image-sequencer/pull/11 that applies my changes on top of @okonek 's. @jywarren can merge both our forks, or @okonek can apply my changes to his fork.

**Comment by jywarren** [comment]
It looked great and I merged it. Great work folks!

**Comment by jywarren** [comment]
So then we can close this one, right? Thanks!!!!

**Comment by jywarren** [comment]
Hi, could you rebase the remaining conflicts and then we'll merge this in? Thanks so much!!!

**Comment by kevinzluo** [comment]
I believe these changes were copied into https://github.com/publiclab/image-sequencer/pull/539 so there is no need to merge this one.

**Comment by jywarren** [comment]
I think you're correct. Thanks!!!!

**Pull request closed without merging**